orttraining/orttraining/training_ops/cuda/optimizer/common.cuh

```diff
@@ -15,11 +15,12 @@ namespace cuda {
 template<typename TLossScale, typename TGradNorm, typename TFinalScale>
 __device__ __forceinline__ TFinalScale _ComputeGradScale(
 const TLossScale* loss_scale,
-const TGradNorm* g_norm) {
+const TGradNorm* g_norm,
+const TFinalScale max_norm) {
 TFinalScale scale = loss_scale != nullptr ? TFinalScale(*loss_scale) : TFinalScale(1.f);
-if (g_norm != nullptr && TFinalScale(*g_norm) > scale) {
+if (g_norm != nullptr && TFinalScale(*g_norm) > TFinalScale(scale * max_norm)) {
     const TFinalScale actual_g_norm = TFinalScale(*g_norm) / scale;
```

Comment:
@wschin do you know why we divide g_norm with scale , got a actual_g_norm ,
 then  scale multiple the actual_g_norm back. This looks interesting, any reason we did this?

Reply:
The actual equation here is to compute a `scalar` which may be used to multiply the gradient passed into optimizer (it's scaled gradient when loss scaling is enabled). If we don't have max gradient norm, it's
```
scalar = loss_scale
```
If max gradient norm is `N`, then we need to further divide the gradient to get `norm=1` gradient. Finally, we multiply the `norm=1` gradient by max norm to make it `norm=max norm` gradient.
```
scalar_for_clipiing = actual_gradient / max_norm
```
and
```
finaly_scalar = scalar * scalar_for_clipping
```